From 3f7eecb62264ea48e9e6ff993ba73fa31657f179 Mon Sep 17 00:00:00 2001
Message-Id: <3f7eecb62264ea48e9e6ff993ba73fa31657f179.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Sagar Kamble <sagar.a.kamble@intel.com>
Date: Wed, 1 Oct 2014 23:40:09 +0530
Subject: [PATCH 304/312] FOR_UPSTREAM [VPG]: drm/i95: Fixing kernel panic due
 to NULL pointer dereference

Following crash caused kernel panic. This happened due to dereferencing of pinned_bo and active_bo that were null.

[ 7235.513665] BUG: unable to handle kernel NULL pointer dereference at 0000000000000088
[ 7235.522464] IP: [<ffffffff815a0896>] i915_capture_error_state+0x946/0x1440
[ 7235.530188] PGD 0
[ 7235.532449] Oops: 0000 1 PREEMPT SMP
[ 7235.536865] Modules linked in: atomisp_css2400b0_v21 videobuf_vmalloc videobuf_core 8723bs(O) mt9m114 cfg80211 hci_uart bluetooth 6lowpan_iphc ov5693 8250_dw ip6table_raw iptable_raw rfkill_gpio bmm050 bmg160 cm32181 bmc150_accel
[ 7235.559809] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G W O 3.14.0-x86_64-01393-gdc3b92d #58
[ 7235.569748] Hardware name: TrekStor SurfTab/ST70408, BIOS ESKU_X64_R_2014_39_5_01_Eng 09/25/2014
[ 7235.579590] task: ffffffff82210480 ti: ffffffff82200000 task.ti: ffffffff82200000
[ 7235.587969] RIP: 0010:[<ffffffff815a0896>] [<ffffffff815a0896>] i915_capture_error_state+0x946/0x1440
[ 7235.598410] RSP: 0000:ffff88003b203c18 EFLAGS: 00010286
[ 7235.604360] RAX: ffff88003c1835c0 RBX: ffff8800370cd000 RCX: ffff88000f3a0400
[ 7235.612350] RDX: ffff88000f3a0430 RSI: 0000000000000000 RDI: ffff880037cc1990
[ 7235.620341] RBP: ffff88003b203ca8 R08: ffffea0000c14650 R09: 0000000000000000
[ 7235.628332] R10: 0000000000000000 R11: 00000000fffff04b R12: ffff88000f3a0400
[ 7235.636324] R13: ffff880037cc0000 R14: ffff880037cc0000 R15: ffff880037c6a800
[ 7235.644317] FS: 0000000000000000(0000) GS:ffff88003b200000(0000) knlGS:0000000000000000
[ 7235.653379] CS: 0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[ 7235.659814] CR2: 0000000000000088 CR3: 000000000220b000 CR4: 00000000001007f0
[ 7235.667802] Stack:
[ 7235.670053] ffff88003b203d00 ffff880037c6a800 ffff88003b203d00 ffff880037cc0000
[ 7235.678368] ffff880037cc1990 ffff880037cc19b8 ffff880037cc1830 ffff880005089c00
[ 7235.686682] ffff8800370cd2a8 ffff8800370cd1f8 ffff880037cc1830 ffff8800370cd000
[ 7235.694996] Call Trace:
[ 7235.697735] <IRQ>
[ 7235.699890] [<ffffffff815a8a3c>] i915_handle_error+0x8c/0x7e0
[ 7235.706657] [<ffffffff8112b134>] ? tick_program_event+0x24/0x30
[ 7235.713396] [<ffffffff810ec337>] ? hrtimer_interrupt+0x137/0x230
[ 7235.720234] [<ffffffff815ba478>] ? __gen6_gt_wait_for_thread_c0.isra.15+0x28/0xa0
[ 7235.728724] [<ffffffff81ad5bf6>] ? _raw_spin_unlock_irqrestore+0x26/0x50
[ 7235.736336] [<ffffffff815bd7d5>] ? vlv_read32+0x95/0x1d0
[ 7235.742395] [<ffffffff815a931d>] i915_hangcheck_hung+0x18d/0x200
[ 7235.749227] [<ffffffff815ab4f9>] i915_hangcheck_sample+0x219/0x2f0
[ 7235.756254] [<ffffffff814798c2>] ? __percpu_counter_sum+0x72/0x80
[ 7235.763188] [<ffffffff815ab2e0>] ? i915_queue_hangcheck+0x90/0x90
[ 7235.770119] [<ffffffff810d1b36>] call_timer_fn+0x36/0x150
[ 7235.776272] [<ffffffff815ab2e0>] ? i915_queue_hangcheck+0x90/0x90
[ 7235.783203] [<ffffffff810d1f94>] run_timer_softirq+0x214/0x2c0
[ 7235.789844] [<ffffffff810cb6b5>] __do_softirq+0x125/0x320
[ 7235.795997] [<ffffffff810cbb95>] irq_exit+0xb5/0xc0
[ 7235.801566] [<ffffffff81ad8d55>] smp_apic_timer_interrupt+0x45/0x60
[ 7235.808689] [<ffffffff81ad79af>] apic_timer_interrupt+0x6f/0x80
[ 7235.815415] <EOI>

For: GMINL-1950
Change-Id: Ifce8c3e56abd51a1e4d15338940ffb8999e3f486
Signed-off-by: Sagar Kamble <sagar.a.kamble@intel.com>
---
 drivers/gpu/drm/i915/i915_gpu_error.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gpu_error.c b/drivers/gpu/drm/i915/i915_gpu_error.c
index 66cf417..95c8ea8 100644
--- a/drivers/gpu/drm/i915/i915_gpu_error.c
+++ b/drivers/gpu/drm/i915/i915_gpu_error.c
@@ -1003,8 +1003,10 @@ static void i915_gem_capture_vm(struct drm_i915_private *dev_priv,
 			capture_pinned_bo(pinned_bo,
 					  error->pinned_bo_count[ndx],
 					  &dev_priv->mm.bound_list);
-	error->active_bo[ndx] = active_bo;
-	error->pinned_bo[ndx] = pinned_bo;
+	if (error->active_bo)
+		error->active_bo[ndx] = active_bo;
+	if (error->pinned_bo)
+		error->pinned_bo[ndx] = pinned_bo;
 }
 
 static void i915_gem_capture_buffers(struct drm_i915_private *dev_priv,
-- 
1.7.9.5

